aten/src/ATen/native/cuda/ReplicationPadding.cu

```diff
@@ -184,26 +311,406 @@ void replication_pad1d_backward_out_cuda_template(
+  }
+  int iheight = input.size(dimh);
+  int iwidth = input.size(dimw);
+  int oheight = iheight + padT + padB;
+  int owidth  = iwidth + padL + padR;
+
+  AT_CHECK(owidth == gradOutput.size(dimw),
+      "gradOutput width unexpected. Expected: ", owidth, ", Got: ",
+      gradOutput.size(dimw));
+  AT_CHECK(oheight == gradOutput.size(dimh),
+      "gradOutput height unexpected. Expected: ", oheight, ", Got: ",
+      gradOutput.size(dimh));
+
+  gradInput.resize_as_(input);
+  gradInput.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      input.type(), "replication_pad2d_backward", [&] {
+
+        auto gradInput_ = gradInput;
+        auto gradOutput_ = gradOutput;
+        if (numInputDims == 3) {
+          gradInput_ = gradInput.reshape({1, gradInput.size(0),
+              gradInput.size(1), gradInput.size(2)});
+          gradOutput_ = gradOutput.reshape({1, gradOutput.size(0),
+              gradOutput.size(1), gradOutput.size(2)});
+        }
+        auto devGradInput = gradInput_.packed_accessor<scalar_t, 4>();
+        auto devGradOutput = gradOutput_.packed_accessor<scalar_t, 4>();
+
+        int outputPlaneSize = devGradOutput.size(2) * devGradOutput.size(3);
+        dim3 gridSize(THCCeilDiv(outputPlaneSize, 256),
+          devGradOutput.size(1),
+          devGradOutput.size(0));
+        dim3 blockSize(outputPlaneSize > 256 ? 256 : outputPlaneSize);
+        replication_pad_backward_kernel <<<gridSize, blockSize, 0,
+        at::cuda::getCurrentCUDAStream()>>>(devGradInput, devGradOutput,
+          padT, padB, padL, padR);
+      }
+  );
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+static inline void shapeCheck3d(
+    const Tensor& input,
+    int pleft, int pright,
+    int ptop, int pbottom,
+    int pfront, int pback) {
+  AT_CHECK(at::cuda::detail::canUse32BitIndexMath(input), 
+      "input tensor must fit into 32-bit index math");
+  int numInputDims = input.dim();
+
+  AT_CHECK(input.numel() && (numInputDims == 4 || numInputDims == 5),
+      "non-empty 4D or 5D (batch mode) tensor expected for input, but got: ", input);
+
+  int planeDim = 0;
+  int dimd = 1;
+  int dimh = 2;
+  int dimw = 3;
+  if (numInputDims == 5) {
+    planeDim++;
+    dimd++;
+    dimh++;
+    dimw++;
+  }
+
+  int numPlanes = input.size(planeDim);
+  int idepth = input.size(dimd);
+  int iheight = input.size(dimh);
+  int iwidth = input.size(dimw);
+  int odepth = idepth + pfront + pback;
+  int oheight = iheight + ptop + pbottom;
+  int owidth  = iwidth + pleft + pright;
+  AT_CHECK(owidth >= 1 || oheight >= 1 || odepth >= 1,
+      "input (D: ", idepth, " H: ", iheight, ", W: ", iwidth,
+      ") is too small."
+      " Calculated output D: ", odepth, " H: ", oheight, " W: ", owidth);
+
+}
+
+static inline void shapeAndGradOutputCheck3d(
+    const Tensor& input,
+    const Tensor& gradOutput,
+    int pleft, int pright,
+    int ptop, int pbottom,
+    int pfront, int pback) {
+  AT_CHECK(at::cuda::detail::canUse32BitIndexMath(input), 
```

Comment:
I'm not the biggest fan of the straight up copy paste of the prelude of this function here. Why not something like optional<Tensor> gradOutput, so that you can test if a gradOutput was passed in or not?